### PR TITLE
More detailed view_adaptor doc

### PR DESCRIPTION
**Pull request by tower120**
By the way, why `empty()` is needed? `equal()` could accept sentinel instead...

### CODE REVIEW

File: doc/index.md [hunk 1]
```diff
@@ -331,6 +331,157 @@ to inject things into the public interface of the iterator:
     };
 ~~~~~~~
 
+From within mixin you can call:
+* `get()` - to access adaptor internals 
+* `base()` - to access adopted iterator
```
Comment (ericniebler) [inline doc/index.md]: adopted?

File: doc/index.md [hunk 2]
```diff
@@ -331,6 +331,157 @@ to inject things into the public interface of the iterator:
...
+
+Iterator/sentinel adaptor may "override" following members:
+~~~~~~~{.cpp}
+    class adaptor : public ranges::adaptor_base
+    {
+        // !For begin_adaptor only!
+        template<typename Rng>
+        constexpr decltype(auto) begin(Rng &rng)
```
Comment (ericniebler) [inline doc/index.md]: s/`decltype(auto)`/`auto`/

File: doc/index.md [hunk 3]
```diff
@@ -331,6 +331,157 @@ to inject things into the public interface of the iterator:
...
+        {
+            return *this.n == that_adapt.n;
+        }
+    };
+
+    adaptor begin_adaptor() const { return {}; }
+    adaptor end_adaptor()   const { return {100}; }
+~~~~~~~
+
+Pay attentention that all data, that you'll store in adaptor - will become part of iterator payload.
```
Comment (ericniebler) [inline doc/index.md]:
```suggestion
Note that all the data that you store in the adaptor will become part of the iterator.
```

File: doc/index.md [hunk 4]
```diff
@@ -331,6 +331,157 @@ to inject things into the public interface of the iterator:
...
+
+If you will not "override" `begin_adaptor()` or/and `end_adaptor()` in your view_adaptor - default ones will be used.
```
Comment (ericniebler) [inline doc/index.md]:
```suggestion
If you will not "override" `begin_adaptor()` or/and `end_adaptor()` in your view_adaptor, default ones will be used.
```

**Comment by ericniebler** [comment]
> By the way, why `empty()` is needed? `equal()` could accept sentinel instead...

I don't remember. 😳

File: doc/index.md (hunk 1 above)
Comment (tower120) [reply doc/index.md]: adapted

File: doc/index.md (hunk 2 above)
Comment (tower120) [reply doc/index.md]: ??

File: doc/index.md (hunk 4 above)
Comment (tower120) [reply doc/index.md]: Applying, but I disagree that commas are better than dashes.

**Comment by tower120** [comment]
> I don't remember. 

I assume, that was needed for the case, when you compared [sentinel with sentinel (?)], **and** [sentinel with iterator]. Probably, with separate `empty()`, dispatching looked more naturally in initial versions...

That all would be just a matter of taste, if not this particular case (not that rare BTW): https://github.com/ericniebler/range-v3/issues/1127

_... Maybe it could be removed... Some day? Refactoring of all views using `empty()` looks trivial..._

File: doc/index.md (hunk 2 above)
Comment (ericniebler) [reply doc/index.md]: Yes, really.

Comment (tower120) [reply doc/index.md]: `??` = don't understand you completely.

----

`s/decltype(auto)/auto/` how this translates to English?

Comment (ericniebler) [reply doc/index.md]: Oh! I'm so sorry. I was speaking sed. /facepalm. Let me try English. Substitute ("s/") `auto` for `decltype(auto)`. In other words:

```suggestion
        constexpr auto begin(Rng &rng)
```

**Comment by ericniebler** [comment]
In all honesty, I'm happier with the design of `view_facade` than I am of `view_adaptor`. I've considered dropping `view_adaptor` entirely and just using facade. Before I do that, I'll probably want to see if could redesign the adaptor so that I'm happier with it.

TL;DR: This all needs thought.

**Comment by tower120** [comment]
>  I've considered dropping view_adaptor entirely and just using facade. 

You need machinery to override iterator/sentinel "customization points". 

I think there is need in `cursor_adaptor` (not the current one), which would have the same interface as cursor (without begin/end/empty/etc.), and would just forward all cursor calls to underlying iterator:

```cpp

template<class BaseIterator, class Derived = void>
struct cursor_adaptor{
    BaseIterator base;

    cursor_adaptor() = default;
    cursor_adaptor(BaseIterator base) : base(base) {}

     // enable_if Derived != void   ( optionally hide base() )
    struct mixin : ranges::basic_mixin<Derived>
    {
        using ranges::basic_mixin::basic_mixin;

        BaseIterator&  base();
    };  

    // forward cursor interface whenever possible
    void next() { 
        ++base;
    }
};

struct my_cursor : cursor_adaptor<BaseIterator, my_cursor>{
    void next() { 
        base += 2;
    }    

    // optional
    struct mixin : my_cursor:::cursor_adaptor::mixin
    {
        // add anything you need.
    };  

};

basic_iterator<my_cursor> my_iter(vec.begin());
```

On top of that you can build everything.
